### PR TITLE
TD 2959 Add new props and stories

### PR DESCRIPTION
**Pull request by m4manjesh**
Closes [TD-2959](https://sce.myjetbrains.com/youtrack/issue/TD-2959/RUI-updates)

## Changes

- `EditLabelDialog` moved to Dialog folder and exported to make it independent component.
- Added two new stories for default and edit.
- Whitespace trimming for name and description handled. Save button only enabled if name has valid values and if description adding and it also has a vlaid value. 
- `selected` and `description` props added for `LabelChip`
- Displaying tick icon if `selected` is true
- Showing Tooltip if  `description` passed
- Stories added for `selected` and `tooltip`
- Tests updated for new props

## UI/UX
@Sowbhagya-ipg 
Figma: https://www.figma.com/proto/0CVK945S1cs1ku0sUXcgv7/VIRTO.SCENE?node-id=9750-45959&t=HEeK6Oiv4MdDIhJA-0&scaling=min-zoom&content-scaling=fixed&page-id=2875%3A43915

EditLabelDialog
Screenshot 2024-08-12 132703
Screenshot 2024-08-12 132049

LabelChip with selected and tooltip
Screenshot 2024-08-12 132150
Screenshot 2024-08-12 132125

## Testing notes

- Check `EditLabelDialog` is visible under `Dialog` folder and works as expected 
- Check callbacks are fired and whitespace trimming works.
- Check `LabelChip` new props shows selected ico and tooltip as expected

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

### CODE REVIEW

File: src/EditLabelDialog/EditLabelDialog.jsx
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:1]: image

Storybook docs are not being populated with any descriptions. As far as I know for JS components these are auto generated based the jsdoc descriptions in the propTypes at the bottom of this file.

File: src/EditLabelDialog/EditLabelDialog.jsx [hunk 1]
```diff
@@ -24,7 +24,7 @@ export default function EditLabelDialog({
   onNew = () => {},
   onEdit = () => {},
   onClose = () => {},
-  labelDialogTitle = "Edit Label",
+  labelDialogTitle = "Add Label",
   label = { _id: "", color: "#005FA8", description: "", name: "" },
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:28]: It's weird for a component to have a default label in this context. Shouldn't label be a **required** property?

File: src/EditLabelDialog/EditLabelDialog.stories.jsx
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:1]: In the story controls, toggling `isOpen` has no effect:

image

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+// Default
+export const Default = {
+  args: {
+    label: {
+      _id: "",
+      color: "#005FA8",
+      description: "",
+      name: ""
+    }
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:50]: You're missing some defaults here e.g. labelDialogTitle has a default in the component of "Add Label"

File: src/EditLabelDialog/EditLabelDialog.jsx [hunk 3]
```diff
@@ -24,7 +24,7 @@ export default function EditLabelDialog({
   onNew = () => {},
   onEdit = () => {},
   onClose = () => {},
-  labelDialogTitle = "Edit Label",
+  labelDialogTitle = "Add Label",
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:27]: A better property name would be `title`. The component name already contains Label and Dialog.

File: src/EditLabelDialog/EditLabelDialog.jsx [hunk 4]
```diff
@@ -108,7 +112,6 @@ export default function EditLabelDialog({
     setName("");
     setDescription("");
     setColor("#005FA8");
-    onClose(event, "save");
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:111]: Why the change in behaviour here?

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+export default {
+  component: EditLabelDialog,
+  title: "Dialog/EditLabelDialog"
+};
+
+const Template = args => {
+  const [open, setOpen] = React.useState(false);
+  const handleClickOpen = () => {
+    setOpen(true);
+  };
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:17]: Rather than trying to handle the state yourself (and hence get out of sync with the storybook controls) we can use the `useArgs` hook from storybook to handle this state.

https://storybook.js.org/docs/addons/addons-api#useargs

See: https://github.com/IPG-Automotive-UK/react-ui/blob/5cb92d7a5eb62220004d8ee849d244560c179339/src/VehicleSelect/VehicleSelect.stories.tsx#L19-L31 for an example usage in RUI.

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  const handleCancel = args => {
+    setOpen(false);
+    action("cancel")(args);
+  };
+  const handleSave = args => {
+    setOpen(false);
+    action("save")(args);
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:24]: `action` is intended to reflect the function name that we are using. In this case `onClose`, `onEdit`, `onNew`. Currently you are showing `cancel` and `save`.

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+export const EditLabel = {
+  args: {
+    label: {
+      _id: "1",
+      color: "#005FA8",
+      description: "This is a description",
+      name: "Label Name"
+    },
+    labelDialogTitle: "Edit Label"
+  },
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:65]: Save some code duplication and build upon the default args.

```suggestion
  args: {
    ...Default.args,
    labelDialogTitle: "Edit Label"
  },
```

File: src/LabelSelector/LabelChip/LabelChip.stories.tsx [hunk 8]
```diff
@@ -79,3 +79,21 @@ export const Small = {
...
+export const Selected = {
+  args: {
+    ...Default.args,
+    selected: true
+  },
+
+  render: Template
+};
+
+export const Tooltip = {
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.stories.tsx:92]: Missing code doc describing the intention of the story. What use case are we showing the developer with these stories.

File: src/LabelSelector/LabelChip/LabelChip.stories.tsx [hunk 9]
```diff
@@ -79,3 +79,21 @@ export const Small = {
 
   render: Template
 };
+
+export const Selected = {
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.stories.tsx:83]: I expected this story to allow me to click on the LabelChip to select/deselect it. I think this is possible with the `onClick` property but it would be good to demonstrate this use case as per the Figma.

File: src/LabelSelector/LabelChip/LabelChip.tsx [hunk 10]
```diff
@@ -12,31 +13,36 @@ export default function LabelChip({
...
-        [`& .${chipClasses.deleteIcon}:hover`]: {
-          color: theme => theme.palette.getContrastText(color)
-        },
-        visibility: visible ? "visible" : "hidden"
-      }}
-      label={<NoWrapTypography variant="inherit">{label}</NoWrapTypography>}
-      size={size}
-      variant={variant}
-    />
+    <Tooltip title={description}>
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.tsx:22]: image
image

The Figma shows that the tooltips should be left aligned to the LabelChip (at least in the Label Selection popover), but they appear centered in Storybook. @Sowbhagya-ipg can you confirm the expectation here?

Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.tsx:22]: The font size for the tooltip in Figma is also different, although we are using the MUI default here which matches tooltips across the rest of our components. What do you want here @Sowbhagya-ipg?

File: src/EditLabelDialog/EditLabelDialog.jsx (hunk 3 above)
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:27]: Why the change in default here? Does this affect any other usages?

File: src/LabelSelector/LabelChip/LabelChip.test.tsx [hunk 11]
```diff
@@ -77,4 +77,31 @@ describe("LabelChip", () => {
...
+    render(<LabelChip label="Label" selected />);
+
+    // get the selected icon
+    const selectedIcon = screen.getByTestId("DoneIcon");
+
+    // check that the selected icon is rendered
+    expect(selectedIcon).toBeInTheDocument();
+  });
+  // renders tooltip on hover of label chip
+  it("renders tooltip on hover", async () => {
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.test.tsx:91]: One of the requirements is that no tooltip is rendered when there isn't a description. Can we add a test that captures this?

**Review by mattcorner — CHANGES_REQUESTED** [review]

File: src/EditLabelDialog/EditLabelDialog.jsx [hunk 12]
```diff
@@ -197,11 +200,12 @@ export default function EditLabelDialog({
...
           color="primary"
           disabled={
-            name.length === 0 ||
+            name.trim().length === 0 ||
             color.length === 0 ||
             !isLabelNameValid ||
             !hasChanged ||
-            !isLabelLengthValid
+            !isLabelLengthValid ||
+            !isDescriptionValid
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx:208]: Me and @Sowbhagya-ipg have some reservations about this logic. Currently I can have a name of "       name" and you pass this on in the `onSave` and `onEdit` callbacks. We should be trimming before returning these callbacks so that while the component shows "     name", the callback receives "name".

The save button should not be disabled when a user has entered a description of "             " as it is not a required field, and it will be trimmed down to an empty description in the callbacks which is valid.

File: src/LabelSelector/LabelChip/LabelChip.tsx (hunk 10 above)
Comment (Sowbhagya-ipg) [reply src/LabelSelector/LabelChip/LabelChip.tsx:22]: Left align it. @m4manjesh

Comment (Sowbhagya-ipg) [reply src/LabelSelector/LabelChip/LabelChip.tsx:22]: I have used the MUI Default too. I wonder why it is not the same. The tooltip font size that I have is 10px.

File: src/EditLabelDialog/EditLabelDialog.jsx (hunk 1 above)
Comment (m4manjesh) [reply src/EditLabelDialog/EditLabelDialog.jsx:28]: I think it should be

File: src/LabelSelector/LabelChip/LabelChip.tsx (hunk 10 above)
Comment (mattcorner) [reply src/LabelSelector/LabelChip/LabelChip.tsx:22]: Looks like we updated the default to be 14px 4+ years ago across all tooltips:

https://github.com/IPG-Automotive-UK/react-ui/blob/a742c9a1635c05fa7a631126dbfad76be1479be6/src/ThemeProvider/ThemeProvider.tsx#L69-L76

Comment (m4manjesh) [reply src/LabelSelector/LabelChip/LabelChip.tsx]: Will wait @Sowbhagya-ipg to comment final decision on this.

File: src/LabelSelector/LabelChip/LabelChip.test.tsx (hunk 11 above)
Comment (m4manjesh) [reply src/LabelSelector/LabelChip/LabelChip.test.tsx:98]: added

File: src/LabelSelector/LabelChip/LabelChip.stories.tsx (hunk 9 above)
Comment (m4manjesh) [reply src/LabelSelector/LabelChip/LabelChip.stories.tsx]: Story updated. Indroduced a new `showIcon` prop to decide to show the icon.

File: src/EditLabelDialog/EditLabelDialog.jsx (hunk 4 above)
Comment (m4manjesh) [reply src/EditLabelDialog/EditLabelDialog.jsx:111]: Missed when editing, added it back

File: src/LabelSelector/LabelChip/LabelChip.tsx (hunk 10 above)
Comment (Sowbhagya-ipg) [reply src/LabelSelector/LabelChip/LabelChip.tsx]: The tooltip with a 10px is too small and 14px is relatively larger. I would suggest it to to be 12px which is lesser than the body font size we have and also readable. @m4manjesh @mattcorner and font weight

File: src/EditLabelDialog/EditLabelDialog.jsx [hunk 13]
```diff
@@ -24,7 +24,7 @@ export default function EditLabelDialog({
   onNew = () => {},
   onEdit = () => {},
   onClose = () => {},
-  labelDialogTitle = "Add Label",
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.jsx]: You need to update any uses of this:

e.g.https://github.com/IPG-Automotive-UK/react-ui/blob/336f7ab5cb9390c410f839a4d8031412f05df1bd/src/LabelSelector/LabelSelector.jsx#L279

image

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 14]
```diff
@@ -4,36 +4,56 @@ import { Button } from "@mui/material";
...
 
 const Template = args => {
-  const [open, setOpen] = React.useState(false);
+  //  useArgs is a hook that returns the current state of the args object
+  const [{ isOpen }, updateArgs] = useArgs();
+
+  // update the args object with the new isOpen value
+  React.useEffect(() => {
+    updateArgs({ isOpen });
+  }, [isOpen, updateArgs]);
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:21]: Do we need this? What case does this handle? I'm confused because isOpen comes from the useArgs hook already.

File: src/EditLabelDialog/EditLabelDialog.stories.jsx [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        onEdit={handleEditLabel}
+        onNew={handleAddLabel}
+      />
+    </>
+  );
+};
+
+// Default
+export const Default = {
+  args: {
```
Comment (mattcorner) [inline src/EditLabelDialog/EditLabelDialog.stories.jsx:64]: You're still missing some default args. I'll list them for you:

```ts
{
options: [],
isOpen: false,
}
```

File: src/LabelSelector/LabelChip/LabelChip.tsx [hunk 16]
```diff
@@ -12,31 +13,37 @@ export default function LabelChip({
...
-      label={<NoWrapTypography variant="inherit">{label}</NoWrapTypography>}
-      size={size}
-      variant={variant}
-    />
+    <Tooltip title={description} placement="bottom-start">
+      <Chip
+        {...props}
+        className="label-chip"
+        clickable={clickable}
+        icon={showIcon && selected ? <DoneIcon color="inherit" /> : undefined}
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.tsx:28]: Whats the use case for selected without icon? This scenario wasn't accounted for in the original ticket so I'm wondering why we've added a third property.

Comment (mattcorner) [reply src/LabelSelector/LabelChip/LabelChip.tsx:28]: Like now I can set `showIcon` to true, but no icon appears until I also set `selected` to true. Suggest we stick with the original `selected` proposal unless there is new information.

**Review by mattcorner — CHANGES_REQUESTED** [review]

File: src/EditLabelDialog/EditLabelDialog.stories.jsx (hunk 14 above)
Comment (m4manjesh) [reply src/EditLabelDialog/EditLabelDialog.stories.jsx:21]: we dont need it

**Review by mattcorner — CHANGES_REQUESTED** [review]

File: src/LabelSelector/LabelChip/LabelChip.test.tsx [hunk 17]
```diff
@@ -77,4 +77,46 @@ describe("LabelChip", () => {
...
+  });
+  // test that the selected icon is not rendered when selected is false
+  it("does not render selected icon", () => {
+    render(<LabelChip label="Label" selected={false} />);
+
+    // there shouldn't be a selected icon
+    expect(screen.queryByTestId("DoneIcon")).not.toBeInTheDocument();
+  });
+  // renders tooltip on hover of label chip
+  it("renders tooltip on hover", async () => {
```
Comment (mattcorner) [inline src/LabelSelector/LabelChip/LabelChip.test.tsx:98]: Your new tests are throwing a bunch of warnings;

https://github.com/IPG-Automotive-UK/react-ui/actions/runs/10370550853/job/28708693334#step:6:237

**Review by mattcorner — CHANGES_REQUESTED** [review]
